File: Codefores Solutions/2023-08-07-Codeforces Round 891 (Div. 3)/E-Power of Points.py
# Link to Problem Statement
# https://codeforces.com/contest/1857/problem/E

from collections import defaultdict as dd

def solve(lst, n, Hash):
  lst.sort(reverse = 1)
  prefix = [0]
  Sum = sum(lst)
  S = Sum
  prefix = 0
    
  for x in range(n):
    t = prefix - (x * lst[x])
    prefix += lst[x]
    t += (n - x) * lst[x] - S
    S -= lst[x]
    Hash[lst[x]] = t + n
  

for _ in range(int(input())):
  
  n = int(input())
  lst = list(map(int, input().split()))
  # Sum = sum(lst)
  
  Sum = sum(lst)
  lst = [(lst[x], x) for x in range(n)]
  Hash = dd(lambda : -1)
  
  # solve(lst.copy(), n, Hash)
  
  lst.sort(reverse = 1)
  prefix = [0]
  S = Sum
  prefix = 0
    
  for x in range(n):
    t = prefix - (x * lst[x][0])
    prefix += lst[x][0]
    t += (n - x) * lst[x][0] - S
    S -= lst[x][0]
    Hash[lst[x]] = t + n
  
  lst.sort(key = lambda x: x[1])
  for x in lst:
    print(Hash[x], end = ' ')
    
  print()

File: Codefores Solutions/2023-08-07-Codeforces Round 891 (Div. 3)/E-Power of Points.cpp
# Link to Problem Statement
# https://codeforces.com/contest/1857/problem/E

#include <bits/stdc++.h>
using namespace std;

int main() 
{
    int t;
    cin >> t;
    
    while (t--)
    {
      
      int n; cin >> n;
      long long prefix = 0, suffix = 0, helper, t;
      vector<int> lst(n), tmp(n);
      map<int, long long> Hash;
      
      for (int x = 0; x < n; x++)
      {
        cin >> lst[x];
        tmp[x] = lst[x];
        suffix += lst[x];
      }
      
      sort(lst.begin(), lst.end(), [&](int x, int y){return x > y;});
      
      // for (int x: lst) cout << x << ' ';
      // cout << '\n';
      
      for (int x = 0; x < n; x++)
      {
        helper = lst[x];
        t = prefix - (x * helper);
        prefix += helper;
        t += (n - x) * helper - suffix;
        suffix -= helper;
        Hash[helper] = t + n;
      }
      
      for (int x: tmp) cout << Hash[x] << ' ';
      cout << '\n';
    }
    
    return 0;
}
